src/librustc_mir/borrow_check/path_utils.rs

```diff
@@ -381,30 +382,72 @@ fn place_element_conflict<'a, 'gcx: 'tcx, 'tcx>(
-                )
-                | (ProjectionElem::ConstantIndex { .. }, ProjectionElem::Subslice { .. })
-                | (ProjectionElem::Subslice { .. }, ProjectionElem::Index(..))
-                | (ProjectionElem::Subslice { .. }, ProjectionElem::ConstantIndex { .. })
-                | (ProjectionElem::Subslice { .. }, ProjectionElem::Subslice { .. }) => {
+                | (ProjectionElem::Subslice { .. }, ProjectionElem::Index(..)) =>
+                {
                     // Array indexes (`a[0]` vs. `a[i]`). These can either be disjoint
                     // (if the indexes differ) or equal (if they are the same), so this
                     // is the recursive case that gives "equal *or* disjoint" its meaning.
-                    //
-                    // Note that by construction, MIR at borrowck can't subdivide
-                    // `Subslice` accesses (e.g. `a[2..3][i]` will never be present) - they
-                    // are only present in slice patterns, and we "merge together" nested
-                    // slice patterns. That means we don't have to think about these. It's
-                    // probably a good idea to assert this somewhere, but I'm too lazy.
-                    //
-                    // FIXME(#8636) we might want to return Disjoint if
-                    // both projections are constant and disjoint.
-                    debug!("place_element_conflict: DISJOINT-OR-EQ-ARRAY");
+                    debug!("place_element_conflict: DISJOINT-OR-EQ-ARRAY-INDEX");
+                    Overlap::EqualOrDisjoint
+                }
+                (ProjectionElem::ConstantIndex { offset: o1, min_length: _, from_end: false },
+                    ProjectionElem::ConstantIndex { offset: o2, min_length: _, from_end: false })
+                | (ProjectionElem::ConstantIndex { offset: o1, min_length: _, from_end: true },
+                    ProjectionElem::ConstantIndex { offset: o2, min_length: _, from_end: true }) =>
+                {
+                    if o1 == o2 {
+                        debug!("place_element_conflict: DISJOINT-OR-EQ-ARRAY-CONSTANT-INDEX");
+                        Overlap::EqualOrDisjoint
+                    } else {
+                        debug!("place_element_conflict: DISJOINT-ARRAY-CONSTANT-INDEX");
+                        Overlap::Disjoint
+                    }
+                }
+                (ProjectionElem::ConstantIndex { offset: offset_from_begin, min_length: min_length1, from_end: false },
+                    ProjectionElem::ConstantIndex { offset: offset_from_end, min_length: min_length2, from_end: true })
+                | (ProjectionElem::ConstantIndex { offset: offset_from_end, min_length: min_length1, from_end: true },
+                   ProjectionElem::ConstantIndex { offset: offset_from_begin, min_length: min_length2, from_end: false }) =>
+                {
+                    let length = max(min_length1, min_length2);
```

Comment:
Comment (this is my attempt):

---

Logic: we are matching a slice with *some* length dynamically; if both of these patterns matched, it must be at least the greater of the two.